compiler/rustc_mir/src/transform/mod.rs

```diff
@@ -390,6 +391,7 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
 
     // The main optimizations that we do on MIR.
     let optimizations: &[&dyn MirPass<'tcx>] = &[
+        &lower_intrinsics::LowerIntrinsics,
```

Comment:
Currently, the optimization is run even with mir-opt-level=0. This means codegen and CTFE can rely on these intrinsics not existing any more. Is that deliberate? For something called "lowering", it makes sense to be able to rely on it, but then it should be moved to `run_post_borrowck_cleanup_passes` IMO.

Reply:
It doesn't run at mir-opt-level=0. The default situation might also change if CTFE is to be run on unoptimized MIR. Always lowering them seems fine to me, modulo concerns already mentioned in the PR description.

Reply:
Are you sure? In https://github.com/rust-lang/rust/pull/79989 I tried `-Zmir-opt-level=0` and it did not make a difference...

Reply:
I see that the pass is only added in `optimizations`, but then CTFE should fail when the `forget` intrinsic is called in https://github.com/rust-lang/rust/pull/79989, and yet I still see this trace:
```
│ │└┐rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
│ │ ├─0ms  INFO rustc_mir::interpret::step // executing bb0
│ │ ├─0ms  INFO rustc_mir::interpret::step _0 = const ()
│ │ ├─0ms  INFO rustc_mir::interpret::step return
│ │ ├─0ms  INFO rustc_mir::interpret::eval_context popping stack frame (returning from function)
│ │┌┘rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
```
Note the lack of anything happening in `mem::forget`. This was with `RUSTC_LOG=rustc_mir::interpret=info rustc +196ca161be8363503d2d2445bab9840f0ce2bba4 /tmp/forget.rs -Zmir-opt-level=0`, on this code:
```rust
fn main() {
    const fn forget_arg_and_return_4(x: Vec<i32>) -> i32 {
        std::mem::forget(x);
        4
    }

    const FOUR_THE_HARD_WAY: i32 = forget_arg_and_return_4(Vec::new());

    assert_eq!(FOUR_THE_HARD_WAY, 4);
}
```

Reply:
Unless you are building std with -Zmir-opt-level=0 it is already too late for it to have an effect for non-local MIR.

Reply:
Oh, right... so we now have a barely tested code path here, namely the intrinsic implementations in CTFE and codegen, which is only hit when std is compiled with mit-opr-level=0. That seems potentially problematic.